File: src/psilox/graphics/Mesh.java
package psilox.graphics;

import static org.lwjgl.opengl.GL11.*;
import static org.lwjgl.opengl.GL15.*;
import static org.lwjgl.opengl.GL20.*;
import static org.lwjgl.opengl.GL30.*;

import java.nio.FloatBuffer;
import java.util.ArrayList;
import java.util.List;

import psilox.utility.Buffers;

public class Mesh {

	private static List<Integer> vaos = new ArrayList<Integer>();
	private static List<Integer> vbos = new ArrayList<Integer>();
	
	private int vaoID;
	private int vertCount;
	
	public Mesh(int vaoID, int vertCount) {
		this.vaoID = vaoID;
		this.vertCount = vertCount;
	}
	
	public int getVaoID() {
		return vaoID;
	}
	
	public int getVertCount() {
		return vertCount;
	}
	
	public static Mesh loadToVao(float[] positions, float[] tcoords) {
		int vaoID = createVao();
		storeDataInAttribList(Shader.VERTEX_ATTRIB, positions, 3);
		storeDataInAttribList(Shader.TCOORD_ATTRIB, tcoords, 2);
		unbindVao();
		return new Mesh(vaoID, positions.length / 3);
	}
	
	private static int createVao() {
		int vaoID = glGenVertexArrays();
		vaos.add(vaoID);
		glBindVertexArray(vaoID);
		return vaoID;
	}
	
	private static void storeDataInAttribList(int attrib, float[] data, int size) {
		int vboID = glGenBuffers();
		vbos.add(vboID);
		glBindBuffer(GL_ARRAY_BUFFER, vboID);
		FloatBuffer buffer = Buffers.createFloatBuffer(data);
		glBufferData(GL_ARRAY_BUFFER, buffer, GL_STATIC_DRAW);
		glVertexAttribPointer(attrib, size, GL_FLOAT, false, 0, 0);
		glBindBuffer(GL_ARRAY_BUFFER, 0);
	}
	
	private static void unbindVao() {
		glBindVertexArray(0);
	}
	
	public void cleanUp() {
		for(int vao : vaos) {
			glDeleteVertexArrays(vao);
		}
		
		for(int vbo : vbos) {
			glDeleteBuffers(vbo);
		}
	}
	
}

File: src/psilox/graphics/Mesh.kt
package psilox.graphics

import glm.vec2.Vec2
import glm.vec3.Vec3
import org.lwjgl.opengl.GL11.*
import org.lwjgl.opengl.GL15.*
import org.lwjgl.opengl.GL20.*
import org.lwjgl.opengl.GL30.*
import psilox.resource.ResourceLibrary
import psilox.resource.VAOResource
import psilox.resource.VBOResource
import psilox.utility.createFloatBuffer
import psilox.utility.toFloatArray

class Mesh private constructor(positions: Array<Vec3>, tcoords: Array<Vec2>) {

    internal val vao: VAOResource
    internal val vboPos: VBOResource
    internal val vboTex: VBOResource
    internal val verts: Int

    init {
        vao = createVAO()
        vboPos = storeDataInAttribList(VERTEX, positions.toFloatArray(), 3)
        vboTex = storeDataInAttribList(TCOORD, tcoords.toFloatArray(), 2)
        verts = positions.size
        unbindVAO()

        with(ResourceLibrary) {
            addReference(vao)
            addReference(vboPos)
            addReference(vboTex)
        }
    }

    protected fun finalize() {
        with(ResourceLibrary) {
            remReference(vao)
            remReference(vboPos)
            remReference(vboTex)
        }
    }

    companion object {

        val VERTEX = 0
        val TCOORD = 1

        private fun createVAO(): VAOResource {
            val vao = VAOResource(glGenVertexArrays())
            glBindVertexArray(vao.glReference)
            return vao
        }

        private fun storeDataInAttribList(attrib: Int, data: FloatArray, size: Int): VBOResource {
            val vbo = VBOResource(glGenBuffers())
            glBindBuffer(GL_ARRAY_BUFFER, vbo.glReference)
            val buffer = createFloatBuffer(data)
            glBufferData(GL_ARRAY_BUFFER, buffer, GL_STATIC_DRAW)
            glVertexAttribPointer(attrib, size, GL_FLOAT, false, 0, 0)
            glBindBuffer(GL_ARRAY_BUFFER, 0)
            return vbo
        }

        private fun unbindVAO() = glBindVertexArray(0)

        fun unitSquare(): Mesh {
            val verts = arrayOf(
                    Vec3(0, 1, 0),
                    Vec3(0, 0, 0),
                    Vec3(1, 0, 0),
                    Vec3(1, 0, 0),
                    Vec3(1, 1, 0),
                    Vec3(0, 1, 0)
            )

            val tcoords = arrayOf(
                    Vec2(0, 0),
                    Vec2(0, 1),
                    Vec2(1, 1),
                    Vec2(1, 1),
                    Vec2(1, 0),
                    Vec2(0, 0)
            )

            return Mesh(verts, tcoords)
        }

    }

}